### PR TITLE
Many vim-like shortcuts for navigating and manipulating blocks

### PR DESCRIPTION
**Pull request by tntmarket**
This introduces a variety of vim like keyboard shortcuts for navigating blocks/pages, opening links, editing and selecting blocks.

It mainly revolves around remembering a "Currently selected block", which is analogous to the cursor in vim.

**Demo:** https://www.youtube.com/watch?v=ZxtT05CjOyo

I believe this addresses #47, #54, and #18, although in a somewhat intrusive way

### Work In Progress

Still need to:

* pull master, resolve conflicts
* organize the commits (they're kind-of a mess)
* organize the code (could share more code existing modules)
* write documentation/instructions where applicable
* add tests where applicable
* figure out how to turn off the feature by default (it's kind-of intrusive)
* revert the temporary commit

### CODE REVIEW

**Comment by Stvad** [comment]
wow, this looks amazing! will look at it closer tonight-ish

**Comment by Stvad** [comment]
some want's from just trying it out for a bit 

- `u` in normal mode should do "undo"
- character hints >> number hints (though this has problem with conflicting with things/etc, so it's not that straightforward)
- enter should probably map to "go to insert mode" too
- would be nice if `l` were to toggle between the page and sidebar (vs having 2 separate shortcuts)
- `shift-o` to create block above
- in queries/mentions - Want hints to be able to expand various levels of "breadcrumbs" (and probably for going to the actual page as-well)
- hints for complex pages are a bit confusing (not sure how to make them better)
- would be awesome if moving blocks were to work in normal mode (though if I understand it correctly it's more of a Roam thing?)


More general Roam toolkit related:
- make delete block work in "normal mode" and map it to `dd`
- vim normal mode SRS shortcuts 😍
- normal mode shortcut to copy block reference/block embed (`y/Y`)

Surfingkeys:
- still want hints from Surfingkeys at times - would be nice to create a config for it that disables overlapping commands, but keeps nav hints working/etc

**Comment by tntmarket** [comment]
Added a commit for undo/redo, and `shift+o` to prepend a line

The hints can be remapped from numbers to keys, maybe we can set the default hint keys to something else closer to home row or something. Maybe there should option to toggle the hints on/off

I was considering extending `h`/`l` to navigate to the left sidebar as well. I personally like it when spatial shortcuts "dead end" at the edge of the window, cause I don't have to worry about "over shooting". But I think a "toggle panel" or "cycle through panels" would be a good

**Comment by Stvad** [comment]
Hmm, undo/redo/shift-o don't seem to work for me for some reason 🤔

**Comment by tntmarket** [comment]
So one strange thing I noticed is that whenever I introduce a new key binding, I have to

1. go to the settings popup
2. click into the newly introduce text input for the new shortcut
3. wait for the green checkmark to appear

In order for the shortcut to start working (this is while I'm doing `yarn run watch`)

**Comment by Stvad** [comment]
> So one strange thing I noticed is that whenever I introduce a new key binding, I have to
> 
>     1. go to the settings popup
> 
>     2. click into the newly introduce text input for the new shortcut
> 
>     3. wait for the green checkmark to appear
> 
> 
> In order for the shortcut to start working (this is while I'm doing `yarn run watch`)

ouch, that seems like a bug in settings, can you create an issue for that?

**Comment by Stvad** [comment]
also as a general remark - this PR is kind of large. if you can think of meaningful way to split it up in several ones to make review easier - would be appreciated. If not it's ok, but it's going to take me a bit to go through it

**Comment by Stvad** [comment]
One weird thing: In the morning I remapped "Select Panel Left" to `Shift+L` and it worked at that time, but it doesn't anymore after some of the updates

**Comment by Stvad** [comment]
btw just fui - key sequences should also work - https://github.com/greena13/react-hotkeys#key-combinations-vs-sequences (you just have to specify them with the space within them)  
so we can stay closer to VIM in a variety of mappings

File: src/ts/features/block-navigation-mode/vim.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        await onPress(getMode())
+        if (updateView) {
+            updateBlockNavigationView()
+        }
+    },
+})
+
+export const imap = (settings: BlockNavigationModeSetting): Shortcut =>
```
Comment (Stvad) [inline src/ts/features/block-navigation-mode/vim.ts:56]: I feel like it should be straightforward to write a function to generate all the functions below based on a given predicate/list of modes

**Comment by tntmarket** [comment]
> also as a general remark - this PR is kind of large. if you can think of meaningful way to split it up in several ones to make review easier - would be appreciated. If not it's ok, but it's going to take me a bit to go through it

For sure, apologies for that. I'll go through the commits and reorganize them :p

**Comment by tntmarket** [comment]
> > So one strange thing I noticed is that whenever I introduce a new key binding, I have to
> > ```
> > 1. go to the settings popup
> > 
> > 2. click into the newly introduce text input for the new shortcut
> > 
> > 3. wait for the green checkmark to appear
> > ```
> > 
> > 
> > In order for the shortcut to start working (this is while I'm doing `yarn run watch`)
> 
> ouch, that seems like a bug in settings, can you create an issue for that?

Created #64

File: src/ts/features/block-navigation-mode/vim.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
+import {Shortcut} from '../../utils/settings'
+import {blurEverything, updateBlockNavigationView} from './blockNavigationView'
+import {isEditing} from '../../utils/dom'
+import {Selectors} from '../../roam/roam-selectors'
+import {delay} from '../../utils/async'
+
+export type Mode = 'INSERT' | 'VISUAL' | 'NORMAL'
```
Comment (Stvad) [inline src/ts/features/block-navigation-mode/vim.ts:7]: should these be Enums? (it's not an assertion but a question :) )

File: src/ts/features/block-navigation-mode/index.ts [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+                    clearBlockNavigationView()
+                }
+            },
+        }),
+        nmap({
+            id: 'undo',
+            key: 'u',
+            label: 'Undo',
+            onPress: async () => {
+                await Keyboard.simulateKey(90, 0, {key: 'z', metaKey: true})
```
Comment (Stvad) [inline src/ts/features/block-navigation-mode/index.ts:228]: why do you have both `90` and `z` here? :) also please add constants for instances where you use this with number

**Comment by Stvad** [comment]
toggle fold does not work in references, unless you hover the mouse on top of the block

**Comment by Stvad** [comment]
Btw, I often have blocks with more then 4 things that can be referenced, what's up with the limit to 4?:)

**Comment by Stvad** [comment]
If I click on the `todo` checkbox - hints within the block disappear. and I need to move focus elsewhere and back before they appear. I suppose it's pertinent to the discussion on whether they should be show all the time or on-demand

**Comment by tntmarket** [comment]
> Btw, I often have blocks with more then 4 things that can be referenced, what's up with the limit to 4?:)

Oh yeah, 4 is too low, I'm thinking something like 10 might be a reasonable upper bound. But then again, the whole hint UX deserves some more consideration

**Comment by tntmarket** [comment]
FYI, heads up that visual mode and insert mode are unreliable for me, sometimes the selection gets stuck, and keys stop working. I'll look more into this.

In the React Hotkey debug logs, I sometimes get into a state where pressing "J" is interpreted as "J+whateverWasLastPressed", even if I'm only pressing "J". Unfocusing/focusing the window usually fixes the problem.

I think this has to do with some race condition related to focusing/unfocusing blocks.

**Comment by Stvad** [comment]
btw added `unmapAllExcept(['f', '/', 'C', '?'], /roamresearch\.com/i);` to Surfingkeys config to get just the hints (will probably add more whitelisting later)

**Comment by Stvad** [comment]
Tab/shift-tab should work in normal mode (probably generally true for anything that you'd expect to work with "native" selection)

**Comment by Stvad** [comment]
Also there should probably be mapping for "jump to last block that is not in linked references"

**Comment by houshuang** [comment]
Just to say that this is absolutely mindblowing... I've been dreaming of something like this! I hate how all the blocks change when you move up/down with a keyboard (especially images etc). I probably even use fewer Markdown links because of that...

This will quickly become something I could not live without. 

Would be very useful with an overview of keyboard shortcuts :) I looked at the source code and tried stuff - worked very well! Love how smooth it feels... 

Would love line numbers (even more fancy is relative line numbers) on the left, and being able to do 10k etc to jump up to a specific line. 

Also hope I can remap Esc to jk keychord as I have in my Vim setup :)

Fantastic work! 🎉🔥

**Comment by classicrob** [comment]
This is absolutely amazing. I'm sure as I'll play around with this I'll have more feedback, but for now the only major thing that is missing for me is unwinding context on a block in a query or a linked reference, or while zoomed in on a block.

**Comment by Stvad** [comment]
`Close Page in Side Bar` does not seem to work for me (it gets invoked as I see the log) but nothing happens

**Comment by Stvad** [comment]
Another great thing to port from Vim that would minimize need for hints: Move cursor to the Top/Middle/Bottom **visible** block (vim analogy would be `H/M/L` correspondingly in visual mode)

**Comment by tntmarket** [comment]
Hey all, thanks for the feedback!

I'm planning to compile a list of bugs/features, decide which ones include in this release, and file issues for the rest. Thinking maybe sometime next week.

**Comment by Stvad** [comment]
Yeah, that sounds good. I as a reviewer vote for small foundational core as a first PR followed by more features based on that separately :)

**Comment by Stvad** [comment]
ah, just noticed it but: arrow keys should work in normal mode

**Comment by classicrob** [comment]
> btw added `unmapAllExcept(['f', '/', 'C', '?'], /roamresearch\.com/i);` to Surfingkeys config to get just the hints (will probably add more whitelisting later)

@Stvad Is it possible to remap a key just for roamresearch? I ended up shifting my up-left-down-right to e-s-d-f, and that conflicts with how I have lowercase hints set up with f. Ideally I'd like to change the surfingkeys command for hints exclusively for this website.

**Comment by tntmarket** [comment]
I realize this is a lot of commits. Once I'm comfortable with where the code is, I'll organize the commits via rebase, and open a new pull request.

File: src/ts/utils/react-hotkeys.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+ * React Hotkeys thinks that the "J" key is still down, which breaks all hotkeys until you
+ * Un-focus the browser window (resetting the key history).
+ *
+ * To workaround this issue, we normalize j/J to be the "same key" using the keycode,
+ * That way, a "J" keydown is stored as a "j" keydown in the key history, so a "j" keyup
+ * will "release" it.
+ *
+ * Related issue: https://github.com/greena13/react-hotkeys/issues/249
+ *
+ * React Hotkeys doesn't have an active maintainer, so it's harder to fix upstream.
```
Comment (Stvad) [inline src/ts/utils/react-hotkeys.ts:38]: should we consider switching to something else?

File: src/ts/features/shortcuts.tsx [hunk 5]
```diff
@@ -1,22 +1,110 @@
...
+    }
+    executingHandler -= 1
+}
+
 export async function updateShortcuts() {
+    const keyMap = await Features.getCurrentKeyMap()
+    const handlers = Features.getShortcutHandlers()
+
+    /**
+     * For some reason, key sequences like 'g g' mess up the other shortcuts
```
Comment (Stvad) [inline src/ts/features/shortcuts.tsx:55]: hm, on the other hand it'd be to great to have sequences work, on the other hand if it requires a lot of custom code, that gives me some hesitation..

**Comment by houshuang** [comment]
The toggleFoldBlock seems to fairly reliable work in the main text, and not work in linked references for me. However, I got it working in linked references at some point. I tried playing around a bit with CSS (to see if the selectors needed to be different) or delay etc, but I couldn't figure out why it was acting differently.

File: src/ts/utils/react-hotkeys.ts (hunk 4 above)
Comment (tntmarket) [reply src/ts/utils/react-hotkeys.ts:38]: I think it's worth researching, I filed https://github.com/roam-unofficial/roam-toolkit/issues/68

**Comment by tntmarket** [comment]
> The toggleFoldBlock seems to fairly reliable work in the main text, and not work in linked references for me. However, I got it working in linked references at some point. I tried playing around a bit with CSS (to see if the selectors needed to be different) or delay etc, but I couldn't figure out why it was acting differently.

Yeah, it's unreliable for me in the linked references too. It works for me if I hover over with the mouse, and then trigger `toggleFoldBlock` though. The issue might be with triggering the mouse hover over the "fold caret button".

**Comment by Stvad** [comment]
When block disappears from the page (when higher level block/page is collapsed, or when you change the date in daily page references). The vim mode looses context i.e. instead of selecting a block nearby or something like that - nothing is selected and you need to explicitly select a new block to get the vim mode back.

**Comment by Stvad** [comment]
> > The toggleFoldBlock seems to fairly reliable work in the main text, and not work in linked references for me. However, I got it working in linked references at some point. I tried playing around a bit with CSS (to see if the selectors needed to be different) or delay etc, but I couldn't figure out why it was acting differently.
> 
> Yeah, it's unreliable for me in the linked references too. It works for me if I hover over with the mouse, and then trigger `toggleFoldBlock` though. The issue might be with triggering the mouse hover over the "fold caret button".

I supposed one option to work around it is to do something like (as an alternative way of doing folding):
* go insert mode
* simulated `cmd-up/down`
* go back to normal mode

File: src/ts/features/block-navigation-mode/index.ts [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+                }
+            },
+        }),
+        nmap({
+            id: 'enterVisualMode',
+            key: 'V',
+            label: 'Enter Visual Mode',
+            onPress: async () => {
+                const block = selectedBlock()
+                if (block) {
```
Comment (Stvad) [inline src/ts/features/block-navigation-mode/index.ts:236]: this is repeated often, would probably makes sense a function like `applyToSelected` that would execute supplied function only if the block is present. Potentially also abstracting away the "returnToNormalMode"

File: src/ts/features/block-navigation-mode/blockNavigation.ts [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+export type Panel = 'MAIN' | 'SIDEBAR'
+
+type State = {
+    mainBlockId: string | null
+    sideBlockId: string | null
+    panel: Panel
+}
+
+export const state: State = {
```
Comment (Stvad) [inline src/ts/features/block-navigation-mode/blockNavigation.ts:12]: I believe this should be a singleton object that'd encapsulate management of the state vs allowing this to be globally writable

File: src/ts/features/block-navigation-mode/blockNavigation.ts [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    }
+}
+
+const blockIdRelative = (block: HTMLElement, blocksToJump: number, panel: Element): string => {
+    const blocks = panel.querySelectorAll(Selectors.block)
+
+    let blockIndex
+    // @ts-ignore NodeList.entries() does exist in Chrome/FF:
+    // https://developer.mozilla.org/en-US/docs/Web/API/NodeList/entries
+    for (let [i, _block] of blocks.entries()) {
```
Comment (Stvad) [inline src/ts/features/block-navigation-mode/blockNavigation.ts:68]: wouldn't using `forEach` which ts recognizes give us the same result?

File: src/ts/roam/roam.ts [hunk 9]
```diff
@@ -39,11 +41,14 @@ export const Roam = {
...
+            await this.activateBlock(element)
+        }
         if (this.getRoamBlockInput()) {
             return Keyboard.pressEsc()
         }
-        return Promise.reject("We're currently not inside roam block")
+        return Promise.reject("We're not inside a block")
     },
 
     async activateBlock(element: HTMLElement) {
```
Comment (Stvad) [inline src/ts/roam/roam.ts:54]: Should reject promise/return null here on else clause I believe

File: src/ts/utils/mouse.ts [hunk 10]
```diff
@@ -2,20 +2,28 @@ import {delay} from './async'
...
-            element.dispatchEvent(getMouseEvent(mouseEventType, buttons))
+            element.dispatchEvent(getMouseEvent(mouseEventType, buttons, shiftKey))
         })
         return delay(delayOverride || this.BASE_DELAY)
     },
-    leftClick(element: HTMLElement, additionalDelay: number = 0) {
-        return this.simulateClick(1, element, additionalDelay)
+    hover(element: HTMLElement, delayOverride: number = 0) {
+        const mouseClickEvents = ['mouseover']
+        mouseClickEvents.forEach(mouseEventType => {
```
Comment (Stvad) [inline src/ts/utils/mouse.ts:14]: the array/loop seem redundant here

File: src/ts/features/block-navigation-mode/index.ts [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+const cutAndGoBackToNormal = async () => {
+    document.execCommand('cut')
+    // Wait for the block to disappear, double check that a block is still selected
+    // Deleting the first line can lead to no previous block existing to select
+    await delay(0)
+    ensureFocusedPanelHasBlockSelected()
+    await returnToNormalMode()
+}
+
+export const config: Feature = {
```
Comment (Stvad) [inline src/ts/features/block-navigation-mode/index.ts:58]: majority of the things here should probably be separate functions in respective modules vs being done inline

**Comment by Stvad** [comment]
My general desire/feeling for the code is that I'd like it to be more "object orienty" in a sense of being composed of small modules/objects that clearly own subsets of functionality. I.e. managing state, managing movement, etc. Would be happy to talk more about it in detail

**Comment by tntmarket** [comment]
Small update: I've been catching up on some work stuff, so haven't had as much time to work on this. Hoping to resume in a week/2

**Comment by Stvad** [comment]
`Escape` does not work while editing page title, when vim mode is enabled

**Comment by Stvad** [comment]
Hmm, if you are to expand breadcrumbs in reference blocks - the next/previous block navigation breaks in a weird way (starts going in loops, on a subset of elements)

btw created a small shortcut for that on a local of your fork :p https://github.com/roam-unofficial/roam-toolkit/tree/block_navigation_mode

**Comment by Stvad** [comment]
also did a bit of restructuring on master recently - merged it into local `block_navigation_mode` branch as-well

**Comment by Stvad** [comment]
something that I do often and would love the hint for is "open references for the topmost page in the sidebar"

File: src/ts/features/block-navigation-mode/blockNavigation.ts (hunk 8 above)
Comment (tntmarket) [reply src/ts/features/block-navigation-mode/blockNavigation.ts:68]: I wanted to short circuit if I found the block early, but I guess I'd have to loop through every block if the last one was selected anyways.

I'll switch to `forEach`

File: src/ts/features/block-navigation-mode/index.ts (hunk 6 above)
Comment (tntmarket) [reply src/ts/features/block-navigation-mode/index.ts:236]: I decided to fallback to selecting the first block, if no block is selected for some reason, which circumvents the null checks.

**Comment by tntmarket** [comment]
Hi everyone, sorry for the wait. I finally cleaned up the code, and broke it into smaller changes so they're easier to review. 

Here are the individual commits, which I hope are independently review-able:

* Absolute Imports - https://github.com/roam-unofficial/roam-toolkit/pull/87
* Various Generic Helpers - https://github.com/tntmarket/roam-toolkit/pull/1
* Hotkey Infrastructure - https://github.com/tntmarket/roam-toolkit/pull/2
* Core Vim Like Functionality - https://github.com/tntmarket/roam-toolkit/pull/3
* Hints - https://github.com/tntmarket/roam-toolkit/pull/4

I'll collect feedback in the other reviews, but push the changes to this mega PR, so the context is cohesively captured in one place.

**Comment by Stvad** [comment]
added comments on 0,1

**Comment by tntmarket** [comment]
Addressed most of the comments in #87 and tntmarket#1, going to try out https://github.com/uzairfarooq/arrive next

**Comment by tntmarket** [comment]
Also added https://github.com/roam-unofficial/roam-toolkit/pull/63, in case anyone wants to try out a horizontal style css theme

**Comment by Stvad** [comment]
@tntmarket regarding `yarn.lock` I removed it mainly because I use npm and GitHub was screaming at me because there were old things in `yarn.lock`. It's probably ok to add it back.  
any good method to keep npm and yarn lock files in sync automagically?

**Comment by nicksettoon** [comment]
I can't contribute, but please don't let this die. vim+roam would be heaven

**Comment by Stvad** [comment]
@nicksettoon I don't think this is in any danger of dying :) @tntmarket and myself are working on bringing this to master. I use it and love it. Absolutely looking forward to pushing it out to the other users!

**Comment by nicksettoon** [comment]
> @nicksettoon I don't think this is in any danger of dying :) @tntmarket and myself are working on bringing this to master. I use it and love it. Absolutely looking forward to pushing it out to the other users!
[My Hero](https://i.gifer.com/2GeW.gif)

**Comment by tntmarket** [comment]
> @tntmarket regarding `yarn.lock` I removed it mainly because I use npm and GitHub was screaming at me because there were old things in `yarn.lock`. It's probably ok to add it back.
> any good method to keep npm and yarn lock files in sync automagically?

Oh actually nvm, I'll use `npm`. For some reason I thought `npm` used `yarn.lock` as it's lockfile lol

**Comment by Stvad** [comment]
btw - I'm curious on why make the PRs internal and not to the master here?

**Comment by tntmarket** [comment]
Oh, I thought it'd make the diffs look more clean, so the later ones wouldn't include changes from the earlier ones

**Comment by Stvad** [comment]
outside of scope of this, but just fyi created https://github.com/roam-unofficial/roam-toolkit/issues/95

**Comment by tntmarket** [comment]
Since we're getting close to releasing this, I created https://github.com/roam-unofficial/roam-toolkit/issues/104 to update the documentation

**Comment by tntmarket** [comment]
Gonna close this, as the functionality has been merged via other PRs and now. Looking forward to the app store release!

**Pull request closed without merging**